rplugin/python3/magma/images.py

```diff
@@ -130,12 +131,166 @@ def add_image(self, path: str, identifier: str, x: int, y: int, width: int, heig
+            self.write_chunked(
+                a='T', # transmit directly to the terminal
+                i=self.id,
+                f=100, # for now, only png
+                v=self.height,
+                s=self.width,
+                C=1,
+                z=10,
+                q=2,
+                data=f.read(),
+            )
+
+    def hide(self):
+        self.nvim.lua.stdout.write(
+            self.serialize_gr_command(
+                i=self.id,
+                a='d', # remove image
+                q=2,
+            )
+        )
+
+
+class Kitty(Canvas):
+    nvim: Nvim
+    images: Dict[str, KittyImage]
+    to_make_visible: Set[str]
+    to_make_invisible: Set[str]
+    visible: Set[str]
+    next_id: int
+
+    def __init__(self, nvim):
+        self.nvim = nvim
+        self.images = {}
+        self.visible = set()
+        self.to_make_visible = set()
+        self.to_make_invisible = set()
+        self.next_id = 0
+        nvim.exec_lua("""
+            local fd = vim.loop.new_pipe(false)
+            fd:open(1)
+            local function write(data)
+                    fd:write(data)
+            end
+
+            stdout = {write = write}
+            """)
+
+    def init(self):
+        return self
+
+    def deinit(self):
+        return
+
+    def log(self,msg):
+        import pynvim
+        self.nvim.api.notify(
+            msg,
+            pynvim.logging.INFO,
+            {'title': "Magma"},
+        )
+
+    def present(self) -> None:
+        # images to both show and hide should be ignored
+        to_work_on = self.to_make_visible.difference(
+            self.to_make_visible.intersection(self.to_make_invisible))
+        self.to_make_invisible.difference_update(self.to_make_visible)
+        for identifier in self.to_make_invisible:
+            self.images[identifier].hide()
+            time.sleep(0.01)
+        for identifier in to_work_on:
+            image = self.images[identifier]
+            def fn(nvim, image):
+                eventignore_save = self.nvim.options["eventignore"]
+                nvim.options["eventignore"] = "all"
+
+                org_position = nvim.current.window.cursor
+                nvim.current.window.cursor = (image.row, image.col)
+
+                image.show()
+
+                nvim.current.window.cursor = org_position
+                nvim.options["eventignore"] = eventignore_save
+            self.nvim.async_call(fn, self.nvim, image)
+        self.visible.update(self.to_make_visible)
+        self.to_make_invisible.clear()
+        self.to_make_visible.clear()
+
+
+    def clear(self):
+        self.log('clear')
+        for identifier in self.visible:
+            self.to_make_invisible.add(identifier)
+        self.visible.clear()
+
+    def add_image(self, path: str, identifier: str, x: int, y: int, width: int, height: int):
+        self.log('add')
+        identifier += f"-{os.getpid()}-{x}-{y}-{width}-{height}"
```

Comment:
Do we need to create a new identifier for every add_image call for the Kitty protocol?
We do this in Ueberzug because over there, if we want to draw the same image in a different position, we need to create a new identifier.
I think this is not the case in Kitty? Not sure though.

Reply:
its probablt safe. I'll change the code and test.